src/test/tags_builder.rs

```diff
@@ -0,0 +1,259 @@
+    EmptyKey,
+    /// An empty set of tags was provided
+    EmptyTags,
+}
+
+impl std::fmt::Display for BuilderError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::EmptyKey => f.write_str("an empty key was provided"),
+            Self::EmptyTags => f.write_str("an empty set of tags was provided"),
+        }
+    }
+}
+
+impl std::error::Error for BuilderError {}
+
+/// A builder for Tags -- this is useful for testing
+///
+/// ```rust
+/// # use twitchchat::irc::{Tags, TagIndices};
+/// # use twitchchat::twitch::{color::RGB, Color};
+/// use twitchchat::test::TagsBuilder;
+///
+/// let (tags_data, tags_indices) = TagsBuilder::new()
+///     .add("color", "#F0F0F0")
+///     .add("display-name", "some-fancy-name")
+///     .add("my-message", "my\nmessage\nspans\nmultiple\nlines")
+///     .build()
+///     .unwrap();
+///
+/// let tags = Tags::from_data_indices(&tags_data, &tags_indices);
+///
+/// let color = tags.get_parsed::<_, Color>("color").unwrap();
+/// assert_eq!(color.rgb, RGB(0xF0, 0xF0, 0xF0));
+///
+/// assert_eq!(tags.get("display-name").unwrap(), "some-fancy-name");
+///
+/// // if the value was escaped, get will returned the escaped string
+/// assert_eq!(tags.get("my-message").unwrap(), r"my\nmessage\nspans\nmultiple\nlines");
+///
+/// // you can get the unescaped value with `get_unescaped`
+/// assert_eq!(tags.get_unescaped("my-message").unwrap(), "my\nmessage\nspans\nmultiple\nlines");
+/// ```
+#[derive(Default, Debug, Clone)]
+pub struct TagsBuilder<'a> {
+    // the spec says 'last' key wins, and the order is irrelevant.
+    // so lets just use a hashmap
+    tags: HashMap<Cow<'a, str>, Cow<'a, str>>,
+}
+
+impl<'a> TagsBuilder<'a> {
+    /// Create a new TagsBuilder
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Add this `key` with this `value` to the builder
+    ///
+    /// # NOTE
+    /// `value` can be empty.
+    /// `key` will replace any previous keys
+    ///
+    pub fn add<K, V>(mut self, key: K, value: V) -> Self
+    where
+        K: Into<Cow<'a, str>>,
+        V: Into<Cow<'a, str>>,
+    {
+        self.tags.insert(key.into(), value.into());
+        self
+    }
+
+    /// Merge these pre-parsed tags into this collection
+    ///
+    /// # NOTE
+    /// This'll override any previously set keys.
+    pub fn merge(mut self, tags: &Tags<'_>) -> Self {
+        self.tags.extend(
+            tags.iter()
+                .map(|(k, v)| (Cow::Owned(k.to_owned()), Cow::Owned(v.to_owned()))),
+        );
+        self
+    }
+
+    /// Build the tags reference string and its indices
+    ///
+    /// # Returns
+    /// This returns a tuple that can be used with [`Tags::from_data_indices`](crate::irc::Tags#from_data_indices)
+    ///
+    /// If any empty keys were found, or no keys at all then an error will be returned,
+    pub fn build(self) -> Result<(MaybeOwned<'static>, TagIndices), BuilderError> {
```

Comment:
From a user perspective this return type is a bit .. lets say weird.

I don't if I should make it more user-friendly -- but Tags are build from a data reference and pre-computed indices.

I could maybe return a simple helper struct which could have a `fn tags(&self) -> Tags<'_>` on it.

Reply:
I implemented this in 920448e